Refactor/toast: Toast 커스텀 훅 정의 및 duration Props 추가

iyu88 opened the pull request:
## 🤠 개요

- #74 
- 기존에 UI적인 부분만 만들어놓았던 Toast 컴포넌트를 실제로 다른 곳에서 사용할 수 있도록 커스텀 훅을 추가 정의했습니다. 

## 💫 설명

- useToast.ts 에 open 상태를 제어할 수 있는 커스텀 훅을 만들었습니다. 

  - WithToastHook 스토리에서 확인할 수 있듯이 커스텀 훅에서 제공하는 함수를 이벤트 핸들러로 등록하여 Toast 를 띄울 수 있습니다. 

  - 이를 위해서 추가적으로 open 과 onChangeOpen 을 Props 에 추가했습니다. 

- Toast 컴포넌트가 자동으로 꺼질 수 있도록 duration Props 를 받습니다. ( default = 3000 = 3초 )

  - 10초동안 떠있는 스토리를 CustomDuration 에서 확인할 수 있습니다. 

- Toast 내부 닫는 버튼 (CloseButton) 에 Button 컴포넌트를 사용하려고 했지만, 수정해야 하는 스타일이 많을 것 같아서
   내부에서 스타일 컴포넌트를 정의해서 사용했습니다. 

- 그 외 사소한 스타일 변경을 했습니다. 

  - Toast 컴포넌트의 배경색을 명시적으로 부여했습니다. 

  - 기존에는 mainColor의 default 값이 빈 문자열이었는데, black 값을 주도록 변경했습니다. 

  - Toast 애니메이션 딜레이가 너무 늦어서 0.1초로 변경하여 이제 더 빨리 동작합니다. 

## 📷 스크린샷 (Optional)

- WithToastHook 스토리 : 3초 후에 사라지고 X 버튼을 누르면 사라집니다.

https://user-images.githubusercontent.com/31645195/227269367-7b8a87cb-b78e-4f21-b154-7e4e6848d220.mov

iyu88 committed:
refactor: CloseButton 태그 및 디자인 변경

iyu88 committed:
feat: useToast 커스텀 훅 생성

Toast 컴포넌트를 사용하는 곳에서 Toast 의 열림 여부를 다루는 Toast 훅을 추가합니다.

iyu88 committed:
feat: Toast 커스텀 훅 관련 props 추가

iyu88 committed:
feat: 커스텀 훅을 사용한 스토리 추가

버튼을 눌렀을 때 Toast 가 열리고 x 버튼을 누르면 Toast 가 닫히는 스토리를 추가합니다.

iyu88 committed:
fix: mainColor 에 기본값을 검정색으로 설정

iyu88 committed:
feat: Toast 를 자동으로 지우는 setTimeout 등록

Toast가 떠 있는 duration을 외부에서 주입받고 해당 시간 이후 useToast의 open을 변경하는 함수를 실행합니다.

iyu88 committed:
feat: duration 관련 스토리 추가

prayinforrain commented on src/components/Toast/index.tsx:
```diff
@@ -97,14 +127,29 @@ const bounce = (vertical: string) => keyframes`
-  border: none;
-  background-color: transparent;
-  cursor: pointer;
+interface ClosebuttonProps {
```
Close**B**uttonProps여야 하지 않나요? 맞춤법 검사 걸려서 코멘트 남깁니당

iyu88 replied:
오타라니.. 
수정했습니다 🥹

prayinforrain reviewed (approved):
useToast 훅은 진짜 너무 멋있네요 bb 사용 방법이 너무 깔끔해요
onChangeOpen은 어떤 함수인가요? open 상태가 변경되었을 때 작동하는 함수(제 추측)가 맞다면 onKeyDown의 형태처럼 onOpenChange 또는 다른 직관적인 이름이었으면 좋을 것 같아요.  
하고 싶은 말은 이게 아니라, hooks를 사용하지 않는 스토리에서 계속 이 함수가 undefined라는 오류가 콘솔에 찍혀서, 필수 props가 맞다면 넣어 주는 게 좋을 것 같아요.

너무너무 멋있어요 이제 자요 우리~~

iyu88 committed:
fix: CloseButtonProps 오타 수정

iyu88 committed:
refactor: onClose 로 핸들러 이름 변경

iyu88 committed:
refactor: useToast 커스텀 훅이 초기값을 받도록 변경

기존에는 Toast 컴포넌트의 open Props 에 default 값으로 false 를 주었습니다.
이러한 방식은 스토리를 확인할 때,
1. open 값을 true 로 전달하여 Toast 컴포넌트를 렌더링해야 하지만
2. 해당 값을 변경하는 핸들러가 없기 때문에 onClose Props 를 비워두게 됩니다.

필수적으로 들어가야 하는 Props 를 전달하지 못하는 문제가 생기기 때문에 커스텀 훅에서 초기값을 받도록 변경합니다.
따라서 아무런 액션을 하지 않고도 Toast 를 보여주고 싶은 상황에서는 useToast(true) 처럼 사용할 수 있습니다.
더불어 useToast 내부에 있는 함수들의 이름도 onVerb 형식으로 변경합니다.

iyu88 commented:
Toast 컴포넌트에서 받는 Props 이름을 onClose 라고 변경했어요. 

닫기 버튼을 누를 때에만 실행할 함수를 받는 위치이기 때문에 Close 라고 이름에 못 박아두었어요.

또한 useToast 내부에 있는 함수들도 onVerb 형태로 이름을 변경했어요. 

나중에 사용하는 곳에서 

`<button onClick={handleToast} />` 보다는 

`<button onClick={onOpenToast} />` 처럼 사용하면 컨벤션에 더 맞을 것 같다고 생각했어요.

<br />

그리고 onClose Props 를 전달하지 않았던 부분은 

open Props 의 defalut 가 false 로 설정되어 있는 상태에서 스토리를 렌더링하기 위해 open 값을 강제로 true로 넣어주었지만

open 값을 변경하는 useState 를 선언하자니 'Toast 렌더링 여부를 결정하는 useToast 가 이미 있는데 굳이?' 라는 생각을 했어요.

가져다가 사용하려고 보니까 useToast 내부 state의 기본값도 false 로 고정되어 있어서 

버튼을 눌러서 띄우는 (= useToast가 적용된) 스토리 이전의 스토리에서는 Toast 의 다른 Props들만 확인할 수 있게 하면 되겠다 싶었어요.

<br />

말씀해주신 부분을 참고하여 useToast 훅이 초기값을 매개변수로 받을 수 있도록 설정함과 동시에

모든 스토리에서 useToast 훅을 사용하는 방식으로 변경하여 

kind, title, message 등의 Props 가 주가 되는 스토리에서 `useToast(true)` 로 초기 open Props 를 true 로 전달하도록 변경했어요. 

<br />

변경 이후에도 오류가 발생하는지 확인해주시면 감사하겠습니다~!

prayinforrain reviewed (commented):
>open Props 의 defalut 가 false 로 설정되어 있는 상태에서 스토리를 렌더링하기 위해 open 값을 강제로 true로 넣어주었지만

실제 usecase에서는 항상 useToast 훅을 같이 사용해야 하기 때문에 발생하지 않는 오류라고 이해하면 될까요?

그리고 핸들러 네이밍 컨벤션 말인데, 지금은 onVerb 형태로 동사만이 명시되어 있지만 onOpenToast처럼 목적어(주어든 목적어든 암튼)가 같이 들어가는 경우가 많을 것 같은데, 여기까지 명확하게 정하면 좋을 것 같아요. 지금 현빈님이 사용하신 형태로 괜찮은지 다들 의견 부탁드려요!

그리고.. 노션에 있는 컨벤션 레포지토리에 이슈로 올려서 pin 해놓으면 어떨까 하는 생각이 들어요. 지금 열기 너무 번거로워요 ㅋㅋ..

iyu88 commented:
현재 useToast 훅을 반드시 사용해야만 한다는 강제성이 없긴 하지만, 

Props를 넘겨주지 않으면 오류가 발생하긴 할테니까 개발 단계에서 useState나 useToast를 써서 값을 전달할 것 같아요.

컨벤션 페이지는 내용을 위키에 클론해놓거나 하면 좋을 것 같긴하네요.

전반적으로 문서 정리를 날 잡아서 해놔야 할 것 같아요..

se030 commented on src/components/Toast/index.tsx:
```diff
@@ -13,64 +14,93 @@ import {
   const { color: themeColor } = theme;
-  const mainColor = kind ? themeColor[kind] : '';
+  const { white, black } = themeColor;
+  const mainColor = kind ? themeColor[kind] : black;
```
이쪽에서 조건부로 확인하는게 아니라 kind default value를 info로 두면 어떨까요?

iyu88 replied:
kind에 default 값을 추가하기 위해서 ToastKind에 `alert` 를 추가했어요.
더불어 color.ts에 black과 동일한 alert 색상을 추가했어요.
이렇게 하니까 mainColor 라는 변수도 더 이상 필요없어지고 기존에 사용하던 위치에서 바로 themeColor[kind]를 사용해요.
확인해보시고 또 의견주세요~

se030 replied:
alert 랑 info랑 다른가요 ???

iyu88 replied:
네! 다릅니다
alert 는 검은색 
info 는 파란색입니다.

se030 commented on src/components/Toast/useToast.ts:
```diff
@@ -0,0 +1,17 @@
+import { useState } from 'react';
+
+const useToast = (isOpen = false) => {
+  const [open, setOpen] = useState<boolean>(isOpen);
```
```tsx
const useToast = (initialState = false) => {
  const [isOpen, setIsOpen] = useState<boolean>(initialState);
```

변수명 이런식으로 쓰는게 더 명확한 것 같아요!

iyu88 replied:
와.. 
open을 isOpen으로 바꾸면 원래 isOpen은 어떡하지? 라고 고민하고 있었는데,
여기에 답이 있었네요
감사합니다 반영할게요~ 👍

se030 commented on src/components/Toast/useToast.ts:
```diff
@@ -0,0 +1,17 @@
+    setOpen(false);
+  };
+
+  return { open, setOpen, onOpenToast, onCloseToast };
```
open, onCloseToast 이렇게 두가지가 Toast 컴포넌트 props로만 사용될 정보라면 아예 toastProps 로 만들어주는거 어때요? 또 외부에서 토스트에 대한 컨트롤을 가지고 싶다면 인터페이스가 아래처럼 있는게 훨씬 사용하기 편하지 않을까요? 예시로 변수명을 좀 바꿔봤어요.

```tsx
  const toastProps = {
    open: isToastOpen,
    onClose: () => setIsToastOpen(false)
  };
  return { setIsToastOpen, toastProps };
```

사용처

```tsx
  const { setIsToastOpen, toastProps } = useToast();

  return (
    <>
      <Button text="열려라 참깨" onClick={ () => setIsToastOpen(true) } />
      <Toast {...args} {...toastProps} />
    </>
  );
```

duration 등의 값들도 핸들러 정의할 때 쓸 수 있는게 좋은가 싶은데 ..ㅎㅎ 제가 말씀드렸던 방식도 급하게 만든거라 구멍이 좀 있을 듯 해서, 어떤 인터페이스를 줬을때 가장 직관적인지, 어떻게 개선해볼 수 있을지 고민을 더 해봐요 ㅎㅎ

iyu88 replied:
제가 이전에 사용해보지 않았던 방식이라 이해하는게 어려웠던 것 같아요.
`() => setIsOpen(true)` 처럼 사용하는 것보다 
토스트 상태를 변경하는 함수 (openToast, closeToast) 를 사용하는게 더 편할 것 같아서 유지했어요.
확인 부탁드려요~

se030 reviewed (commented):
고생하셨어요 👍🏻

leesunmin1231 commented on src/components/Toast/index.tsx:
```diff
@@ -13,64 +14,93 @@ import {
+            </Typography>
+            <Typography variant="desc">{message}</Typography>
+          </Flexbox>
+          <CloseButton mainColor={mainColor} onClick={onClose}>
+            <MdOutlineCancel size={CLOSE_ICON_SIZE} color={mainColor} />
```
여기 CloseButton props는 mainColor로 하신 이유가 있나요? 따로 color props가 있는게 아니라면 color로 통일하는게 좋을 거 같아요!!

leesunmin1231 commented on src/components/Toast/Toast.stories.tsx:
```diff
@@ -105,3 +112,31 @@ Error.args = {
 };
+
+const ToastHookTemplate: ComponentStory<typeof Toast> = (args) => {
+  const { open, onOpenToast, onCloseToast } = useToast();
+
+  return (
+    <>
+      <Button text="열려라 참깨" onClick={onOpenToast} />
+      <Toast {...args} open={open} onClose={onCloseToast} />
```
이건 개인적인 의견인데, 어차피 토스트 메시지를 사용하는 목적이 알림 메시지를 띄우기 위함이니까, 굳이 onClose handler를 바깥에서 사용할 필요가 없어보여요. 바깥에서는 open handler만 사용하고 토스트 컴포넌트 내부에서만 close 동작을 핸들링 해주는건 어때요?

iyu88 replied:
Toast를 여는 state를 외부에서 props로 받는 형태여서 
어쩔 수 없이 해당 값에 대한 setter 함수까지 props로 전달받고 있어요.
더 세련된 방법도 있을 것 같은데 지금 구조로는 useToast 훅과 함께 사용하는 쪽이 좋을 것 같아요.
제가 잘못 이해했거나 추가적인 의견 있으면 말씀해주세요..!!

iyu88 committed:
refactor: useToast 내부 반환값 수정

useToast 에서 사용하는 boolean 상태에 대한 이름을 isOpen 으로 변경합니다.
커스텀 훅이 반환하는 함수명도 이벤트 핸들러 형식이 아닌 일반 함수처럼 변경합니다.
Toast에서 필요한 값을 Object로 묶어서 반환합니다.

iyu88 committed:
refactor: CloseButton Props 이름 수정

iyu88 committed:
refactor: Toast 기본 kind 를 alert 로 설정

kind props 에 default 값을 넣어주기 위해서 ToastKind 타입alert 값을 추가했습니다.
color.ts 에 alert 색상을 추가했습니다.

iyu88 committed:
feat: Alert 스토리 추가 및 스토리 순서 변경
